protocols/imap/src/main/java/org/apache/james/imap/processor/SetAnnotationProcessor.java

```diff
@@ -0,0 +1,78 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.processor;
+
+import java.util.List;
+
+import org.apache.james.imap.api.ImapCommand;
+import org.apache.james.imap.api.ImapConstants;
+import org.apache.james.imap.api.ImapSessionUtils;
+import org.apache.james.imap.api.display.HumanReadableText;
+import org.apache.james.imap.api.message.response.StatusResponse;
+import org.apache.james.imap.api.message.response.StatusResponseFactory;
+import org.apache.james.imap.api.process.ImapProcessor;
+import org.apache.james.imap.api.process.ImapSession;
+import org.apache.james.imap.message.request.SetAnnotationRequest;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.MailboxNotFoundException;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.slf4j.Logger;
+
+import com.google.common.collect.ImmutableList;
+
+public class SetAnnotationProcessor extends AbstractMailboxProcessor<SetAnnotationRequest> implements CapabilityImplementingProcessor {
+
+    public SetAnnotationProcessor(ImapProcessor next, MailboxManager mailboxManager, StatusResponseFactory factory) {
+        super(SetAnnotationRequest.class, next, mailboxManager, factory);
+    }
+
+    public List<String> getImplementedCapabilities(ImapSession session) {
+        return ImmutableList.of(ImapConstants.SUPPORTS_ANNOTATION);
+    }
+
+    protected void doProcess(SetAnnotationRequest message, ImapSession session, String tag, ImapCommand command,
+            Responder responder) {
+        final MailboxManager mailboxManager = getMailboxManager();
+        final MailboxSession mailboxSession = ImapSessionUtils.getMailboxSession(session);
+        final String mailboxName = message.getMailboxName();
+        try {
+            MailboxPath mailboxPath = buildFullPath(session, mailboxName);
+
+            mailboxManager.updateAnnotations(mailboxPath, mailboxSession, message.getMailboxAnnotations());
+
+            okComplete(command, tag, responder);
+        } catch (MailboxNotFoundException e) {
+            log(session, command, mailboxName, e);
+            no(command, tag, responder, HumanReadableText.FAILURE_NO_SUCH_MAILBOX, StatusResponse.ResponseCode.tryCreate());
+        } catch (MailboxException e) {
+            log(session, command, mailboxName, e);
+            no(command, tag, responder, HumanReadableText.GENERIC_FAILURE_DURING_PROCESSING);
+        }
+    }
+
+    private void log(ImapSession session, ImapCommand command, final String mailboxName, MailboxException e) {
+        Logger log = session.getLog();
```

Comment:
session.getLog() returns a org.slf4j.Logger.

Info enabled is embedded in the state of this logger.

So if the logger is able to tell you "I do not have info enabled" why would it log to info anyway ? If it is set  "org.apache.james.imap < INFO" then session.getLog().info(...) will not produce any output.

The only advantage I can see to do this check manually is to avoid arguments computation if not needed. It sounds like early optimisation to me.